joomla/joomla-cms
# [4.1] Add inline editing for articles

### Summary of Changes
This PR adds frontend inline editing functionality to the default single article view (and possibly more modules in the future). 
When hovering over editable elements, a border indicates the ability to change them inline. This aims for quick fixes in case of mispelled titles/words and removes the need to open the dedicated edit page and do the changes in there. The dedicated edit page still persists for other changes like permissions etc.
Right now the PR only targets the default article view but with a more abstract implementation and the use of the HTMLHelper Class the aim for the future would be that adding the HTMLHelper call with a few changes to the form makes possibly every module inline editable.
In the backend there is now an option which allows to enable/disable the inline editing functionality. Right now it is located within the article tab but may be moved to the global options if the inline editing is available in more places. Further the inline editing feature is only accessible to users who are allowed to edit.


### Testing Instructions
1. Checkout this branch
2. run `node build.js --compile-js`
3. visit an article page while logged out
4. try to hover over the title or the text of the article (nothing should happen)
5. log into an account which is able to edit
6. hover over the title or the article text -> a border should appear which indicates the ability to edit
7. click into the field and edit something
8. click somewhere out of the field, the article should be saved with the new changes
9. visit another site and come back, the changes should be persistent
10. go to the backend, find the inline editing option and disable it
11. inline editing should now not be functional anymore


### Actual result BEFORE applying this Pull Request
####  No inline editing features whatsoever
![image](https://user-images.githubusercontent.com/46525639/91170245-e525b700-e6d8-11ea-98fe-4dd3d1dfb702.png)
####  No enable/disable inline editing switch
![image](https://user-images.githubusercontent.com/46525639/91170579-62512c00-e6d9-11ea-9525-0086f1450f23.png)


### Expected result AFTER applying this Pull Request
####  Inline disable/enable switch is present
![image](https://user-images.githubusercontent.com/46525639/91170920-e60b1880-e6d9-11ea-8cc6-6f98fc57d75d.png)
####  On hovering over editable elements the border indicates inline editing
![image](https://user-images.githubusercontent.com/46525639/91171011-10f56c80-e6da-11ea-86fa-42669ba14929.png)
####  On clicking on editable elements the editor is activated and changes can be made
![image](https://user-images.githubusercontent.com/46525639/91171123-400bde00-e6da-11ea-909b-eae88683aa7c.png)
#### On leaving the field the changes are saved
![image](https://user-images.githubusercontent.com/46525639/91171173-5ca81600-e6da-11ea-9ad0-a49f44be83ea.png)



Edit:
### Outlook and further changes
As pointed out there is still work that needs to be done in order to achieve an inline editing functionality that is suitable for several modules in Joomla. For further implementation I'll quickly outline the current structure and possible approaches.
The ultimate goal would be to achieve inline editing funtionality by calling the HTMLHelperClass's function for inline editing and assign classes to the elements that should have the inline editing functionality.
Right now there are a few things that need to be changed in order to achieve this goal:

1. find solution to currently used form
2. remove all hard coded elements 
3. data-handling
4. TinyMCE as the standard inline editor
5. user feedback

1: 
To save the edits which were made in the inline editor there is currently a form used which works similar to the form of the dedicated edit page. Two issues stem from this approach: The redirection which is currently prevented by redirecting to a dummy iframe on the same page and the construction of the missing form arguments via the js file that elsewise would miss and cause the save function to fail. It needs to be decided if this form is actually needed or a simple presentation is sufficient and the saving could be achieved with an AJAX request instead of submitting a form.
2: 
As pointed out before the JS file constructs the missing arguments so that the controller accepts the submitted form. The ultimate goal for the inline editing would be that at no point in code one could tell for which item the inline editing is used right now. This could be achieved with a greater level of abstraction.
3:
Additioal implementation is needed with the data handling as well. There needs to be a concept which regulates how data is collected, sent and extracted + saved that implements the level of abstraction that is desired, maybe the data class can be helpful for that.
4:
We may think that TinyMCE is a good choice as an inline editor but ultimately the user should get to chose what editor should be used. It just needs to be clear that right now a lot of the functionality is carried by TinyMCE.
If there should be several inline editors and a plain text editor in the case of no "smart" editor there are several changes that need to be made (test which editor is available, copy functionalities of TinyMCE and provide them to non editor views, etc)
5:
It may be useful to add any kind of user response to make clear if the inline changes are saved or not. Common message alerts should be avoided to prevent making the quick edits seem "heavy" regarding to usability. Working with a little "saved" an "unsaved" sign or a colored dot might be a choice.

Ultimatedy we'd like to point out that this is actually not a finished product. We began coding for Joomla only a few weeks ago but are glad to introduce a proof of concept of an actual inline editing feature for the whole Joomla! CMS. If you want to pick up the next steps of implementation you're very welcome. If there are any questions, suggestions or tips please join the discussion below.
